### ISSUE TITLE
Adding support for privacy level in string interpolation

### ISSUE THREAD

**Issue by Frizlab** [apple/swift-log#204]
On macOS, the (Apple’s) Logger object uses a custom `OSLogMessage` object which allows it to have a custom string interpolation with support for format arguments privacy levels.

Would it be possible to add support for said string interpolation so we could finally have an efficient and fully feature-compatible OSLog `Logger` for `swift-log`?

Thanks!

**Comment by shaps80** [comment]
While not exactly what you're asking for, I have written something like this and mentioned it on another issue #208 to see if this is appropriate for this package. If not, it's still usable from my own package. Feel free to use.

**Comment by tomerd** [comment]
That is a good idea @nashysolutions. would you like to make a PR with proposed language? I suspect we may also want to consider a new metadata type for sensitive data that the log backend can reduct

**Comment by Mordil** [comment]
We've implemented a similar mechanism in our local code - is this something that might be generally useful?

```
extension Logger.MetadataValue {
  /// The different styles of privacy masks that can be applied to a metadata value.
  internal enum PrivacyMask {
    /// The value will be hashed and converted into an 8 character hex representation.
    case hash
    /// Replaces the value with `*****`.
    case redact
  }

  /// Replaces the value with `*****`.
  ///
  /// This is great for when the actual value is not of importance and should be masked when not debugging.
  ///
  /// For example:
  ///
  ///     let logger: Logger = ...
  ///     logger.info(
  ///       "replaced value with mask pattern",
  ///       metadata: [
  ///         "password": .mask(user.password)
  ///       ]
  ///     )
  public static func mask(_ value: Any) -> Logger.MetadataValue {
    return .stringConvertible(MaskedValue(value, mask: .redact))
  }

  /// The value will be hashed and converted into an 8 character hex representation.
  ///
  /// This is great for when identity of value between individual logs is important, but the actual value should still be masked.
  ///
  /// For example:
  ///
  ///     let logger: Logger = ...
  ///     logger.info(
  ///       "obfuscated value",
  ///       metadata: [
  ///         "ip_address": .hash(user.ipAddress)
  ///       ]
  ///     )
  public static func hash(_ value: Any) -> Logger.MetadataValue {
    return .stringConvertible(MaskedValue(value, mask: .hash))
  }
}

internal struct MaskedValue: CustomStringConvertible {
  internal let underlying: Any

  private let mask: Logger.MetadataValue.PrivacyMask

  internal var description: String {
    switch mask {
    case .hash: return self.hash("\(self.underlying)")
    case .redact: return "*****"
    }
  }

  fileprivate init(_ value: Any, mask: Logger.MetadataValue.PrivacyMask) {
    self.underlying = value
    self.mask = mask
  }

  private func hash(_ value: String) -> String {
    var hasher = Hasher()
    hasher.combine(value)
    return "\(String(format: "%x", hasher.finalize()))"
  }
}
```

**Comment by tomerd** [comment]
this seems interesting to me. @ktoso? @weissi?

**Comment by Mordil** [comment]
Here's an example of the `LogHandler` transforming the metadata

```swift

switch metadataValue {
case let .stringConvertible(maskedValue as MaskedValue) where self.logLevel <= .debug:
      return "\(maskedValue.underlying)"

default: return "\(metadataValue)"
}
```

**Comment by shaps80** [comment]
@Mordil @tomerd @Frizlab I actually open sourced exactly this some time ago. It was discussed with the team here on Swift Forums but was deemed inappropriate for this library.

So I just did my own version that adds the interpolation for similar convenience. It's been used in several projects since its release some time ago and I haven't had any reported issues.

> It obviously doesn't have the same performance characteristics in production builds, since OSLog only executes its closures when logs are being pulled into Console.app or Xcode. It uses `print` which most other logging libraries do, so I don't see this as being an issue. Just full disclosure 👍

[https://github.com/shaps80/Logging](https://github.com/shaps80/Logging)

**Comment by Mordil** [comment]
@shaps80 Do you have links to the discussions where it was determined this was inappropriate for the library to offer directly?

**Comment by shaps80** [comment]
Yeah, https://forums.swift.org/t/swiftlog-stringinterpolation/52313/3
You actually were one of the last comments agreeing to this hahaha.

**Comment by weissi** [comment]
@tomerd For apps I think adding privacy works just fine. For libraries it's IMHO not possible to sensibly annotate. Take a URL, some apps have sensitive URLs, other apps don't but the HTTP client library couldn't know how it's used.

My personal opinion is: If you need privacy, then
- stick everything that's not constant into metadata (e.g. `logger.info("user logged in", metadata: ["user-name": "\(userName)", "request-uuid": "\(request.uuid)"])`)
- in your LogHandler, redact _every_ metadata value that isn't specifically allow listed

So the above message (`logger.info("user logged in", metadata: ["user-name": "\(userName)", "request-uuid": "\(request.uuid)"])`) should IMHO be logged as

```
user logged in [user-name: <redacted>, request-uuid: 1F4A3360-AEEC-48A4-92E4-60DA38823004]
```

because an IMHO sensible LogHandler configuration would (for a hypothetical SuperCoolPrivacyLogHandler) could look something like

```
SuperCoolPrivacyLogHandler(
    label: ...,
    configuration: SuperCoolPrivacyLogHandlerConfiguration(
        nonRedactedMetadataKeysList: ["request-uuid", "error", "timestamp", "count", "ahc-error", "ahc-connection-id", ...]
    )
)
```

The reason is that IMHO only the _application_ can know what has PII and what doesn't. If the HTTP client is only used for requests that never contain PII, then it would be safe to add `ahc-url` or whatever into the `nonRedactedMetadataKeysList` but for many other apps that wouldn't be acceptable at all.

**Comment by tomerd** [comment]
one thing to keep in mind is that the application is also a "producer" log events. its makes perfect sense to me that libraries do not force PII decision on the app, but its nice for the apps to be able to emit ones taking into account PII considerations. that said, I can see how this can be handled in a sophisticated enough handler / backend with nice configuration for this kind of settings

**Comment by weissi** [comment]
> one thing to keep in mind is that the application is also a "producer" log events. its makes perfect sense to me that libraries do not force PII decision on the app, but its nice for the apps to be able to emit ones taking into account PII considerations. that said, I can see how this can be handled in a sophisticated enough handler / backend with nice configuration for this kind of settings

Totally agree with everything. I guess I was trying to say doing something that adds privacy for the app only but ignores the libraries isn't any good. And yes, sophisticated LogHandlers are the way. I wish there was a properly configurable and [featureful "logging middle-end"](https://forums.swift.org/t/thoughts-on-why-does-swiftlog-not-have-multiple-log-destinations-formatters-colouring-some-other-feature/37507) that would take care of these things once and for all.

**Comment by kukushechkin** [comment]
SLG-004 is now accepted and implemented in #453. While it is not direct _privacy attributes_ support, it adds support for generic metadata value attributes and an application can implement a custom attribute, that could be handled in an application-specific log handler.

We would like to keep specific attributes out of `swift-log`, so we would not add the privacy attribute directly to this package.

**Comment by Frizlab** [comment]
I was thinking of this issue when I saw the PR you mentioned being merged 😊
It’s actually a very good new feature, I was very happy to see it land.
